packages/test/test-end-to-end-tests/src/test/gc/newFluidObjectVisibility.ts

```diff
@@ -0,0 +1,361 @@
+ * visible DDS.
+ */
+describeNoCompat.only("New Fluid objects visibility", (getTestObjectProvider) => {
+    let provider: ITestObjectProvider;
+    let container1: IContainer;
+    let containerRuntime1: IContainerRuntime;
+    let dataObject1: ITestDataObject;
+
+    async function requestTestObjectWithoutWait(router: IFluidRouter, id: string): Promise<ITestDataObject> {
+        const request: IRequest = { url: id, headers: { wait: false } };
+        return requestFluidObject(router, request);
+    }
+
+    /**
+     * Creates a non-root data object and validates that it is not visible from the root of the container.
+     */
+    async function createNonRootDataObject(
+        container: IContainer,
+        containerRuntime: IContainerRuntime,
+    ): Promise<ITestDataObject> {
+        const dataStore = await containerRuntime.createDataStore(TestDataObjectType);
+        const dataObject = await requestTestObjectWithoutWait(dataStore, "");
+        // Non-root data stores are not visible (unreachable) from the root unless their handles are stored in a
+        // visible DDS.
+        await assert.rejects(requestTestObjectWithoutWait(container,dataObject._context.id),
+            "Non root data object must not be visible from root after creation",
+        );
+        return dataObject;
+    }
+
+    /**
+     * Creates a root data object and validates that it is visible from the root of the container.
+     */
+    async function createRootDataObject(
+        container: IContainer,
+        containerRuntime: IContainerRuntime,
+        rootDataStoreId: string,
+    ): Promise<ITestDataObject> {
+        const dataStore = await containerRuntime.createRootDataStore(TestDataObjectType, rootDataStoreId);
+        const dataObject = await requestTestObjectWithoutWait(dataStore, "");
+        // Non-root data stores are visible (reachable) from the root as soon as they are created.
+        await assert.doesNotReject(requestTestObjectWithoutWait(container, dataObject._context.id),
+            "Root data object must be visible from root after creation",
+        );
+        return dataObject;
+    }
+
+    async function getAndValidateDataObject(
+        fromDataObject: ITestDataObject,
+        key: string,
+        container: IContainer,
+    ): Promise<ITestDataObject> {
+        const dataObjectHandle = fromDataObject._root.get<IFluidHandle<ITestDataObject>>(key);
+        assert(dataObjectHandle !== undefined, `Data object handle for key ${key} not found`);
+        const dataObject = await dataObjectHandle.get();
+        await assert.doesNotReject(requestTestObjectWithoutWait(container, dataObject._context.id),
+            `Data object for key ${key} must be visible`);
+        return dataObject;
+    }
+
+    async function ensureContainerConnected(container: Container): Promise<void> {
+        if (!container.connected) {
+            return new Promise((resolve) => container.once("connected", () => resolve()));
```

Comment:
i have a utility, timeoutPromise which will automatically timeout if the promise does resolve, there are a bunch of example usages, it helps to prevent test hangs which are hard to debug